c10/core/SafePyObject.h

```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include <c10/core/impl/PyInterpreter.h>
+#include <c10/macros/Macros.h>
+#include <c10/util/python_stub.h>
+
+namespace c10 {
+
+// This is an safe owning holder for a PyObject, akin to pybind11's
+// py::object, with two major differences:
+//
+//  - It is in c10/core; i.e., you can use this type in contexts where
+//    you do not have a libpython dependency
+//
+//  - It is multi-interpreter safe (ala torchdeploy); when you fetch
+//    the underlying PyObject* you are required to specify what the current
+//    interpreter context is and we will check that you match it.
+//
+// It is INVALID to store a reference to a Tensor object in this way;
+// you should just use TensorImpl directly in that case!
+struct C10_API SafePyObject {
+  // Steals a reference to type_object
```

Comment:
do you mean `data`?

Reply:
Is the steal semantic the same as `py::object`? From https://github.com/pybind/pybind11/issues/1201, it doesn't look like it? Should that be a third major difference listed above?

Reply:
py::object just doesn't accept a bare PyObject pointer at all. I suppose if you wanted to be safe we could add explicit borrow/steal APIs...